Code/Framework/AzCore/AzCore/DOM/DomPrefixTree.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/DOM/DomPath.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/containers/stack.h>
+#include <AzCore/std/optional.h>
+
+namespace AZ::Dom
+{
+    //! Specifies how a path matches against a DomPrefixTree
+    enum class PrefixTreeMatch
+    {
+        //! Only an exact path will match.
+        //! For the path "/foo/bar" only "/foo/bar" will match while
+        //! "/foo" and "/foo/bar/baz" will not.
+        ExactPath,
+        //! The path, and any of its subpaths, will match.
+        //! For the path "/foo/bar" both "/foo/bar" and any subpaths like
+        //! "/foo/bar/0" will match, while "/foo" and orthogonal paths like
+        //! "/bar" will not
+        PathAndSubpaths,
+        //! Any of the path's subpaths will match, excepting the path itself.
+        //! For the path "/foo/bar", "/foo/bar" will not match but "/foo/bar/0"
+        //! will.
+        SubpathsOnly,
+    };
+
+    //! A prefix tree that maps DOM paths to some arbitrary value.
+    template<class T>
+    class DomPrefixTree
+    {
+    public:
+        DomPrefixTree() = default;
+        DomPrefixTree(const DomPrefixTree&) = default;
+        DomPrefixTree(DomPrefixTree&&) = default;
+        explicit DomPrefixTree(AZStd::initializer_list<AZStd::pair<Path, T>> init);
```

Comment:
You could also have a constructor accept that accepts an input_range.

```suggestion
        template<class Range, class = enable_if_t<AZStd::input_range<Range> && AZStd::convertible_to<AZStd::tuple_element_t<0, AZStd::range_value_t<Range>>, Path> && AZStd::convertible_to<AZStd::tuple_element_t<1, AZStd::range_value_t<Range>>, T> >>
        explicit DomPrefixTree(Range range);
```

Reply:
Interesting! I'm not terribly familiar with the day-to-day usage of the `range` concept yet - presumably this would let any `pair-like` iterable that provides `Path, T` entries to be a ctor parameter? Is that a general idiom we should try to carry forward as opposed to accepting `initializer_list`?

Reply:
Well the initializer_list constructor is still needed for for when passing multiple elements using the brace syntax.
i.e
```
DomPrefixTree newTree{ Pair{}, Pair{}, Pair{} };
```

You are correct, that the paradigm here is to accept a type that models the range concept and has at least two elements that can be unpacked using `AZStd::get<0>(range[i])` and `AZStd::get<1>(range[i])`, where "range[i]" represents a single element of the range.
This is how tuple like types opt-in to the structured binding syntax: https://en.cppreference.com/w/cpp/language/structured_binding#Case_2:_binding_a_tuple-like_type

Reply:
Right on - I've gone ahead and added a range overload. I also realized when testing the initializer list ctor that I'm not actually doing perfect forwarding in a couple places, fixed those up as well w/ tests.

Reply:
Oh about that ...
An initializer_list cannot be perfect forwarded.
It's iterator type is [const T*](https://en.cppreference.com/w/cpp/utility/initializer_list#Member_types).

Only a copy can be made from it's elements, as an initializer_list doesn't actually own the elements within it.
It is only a view of the those elements

Reply:
Yep that's actually how I noticed it! Perfect forwarding should have meant `SetValue` deduced a const lvalue from the `initializer_list` parameters, but instead it failed to compile because I was actually just accepting an explicit `T&&` rvalue which it couldn't promote to.